policy/lib/bundles.rego

```diff
@@ -32,9 +32,23 @@ out_of_date_task_bundle(tasks) := {task |
 	ref := image.parse(bundle(task))
 	collection := _collection(ref)
 
-	some match_index, out_of_date in collection
-	is_equal(out_of_date, ref)
+	some match_index, match_bundle in collection
+	is_equal(match_bundle, ref)
+
+	# the reference is not the latest
 	match_index > 0
+
+	# all tasks that are newer than the given task
+	some b in array.slice(collection, 0, match_index)
+
+	# is there a newer task for the same version (identified by tag)
+	b.tag == match_bundle.tag
+
+	# it is not the same reference
+	not is_equal(b, ref)
+
+	# and it is newer
+	time.parse_rfc3339_ns(b.effective_on) > time.parse_rfc3339_ns(match_bundle.effective_on)
 }
 
 # Returns a subset of tasks that do not use an acceptable bundle reference.
```

Comment:
Do we need to add some special handling in `unacceptable_task_bundle` and `is_acceptable` as well? If there's enough churn for tag A, tag B may get buried down the list and trigger a violation when used.

Consider making the changes in `_collection` so it covers the different cases:
```rego
_collection(ref) := items {
	full_collection := data["task-bundles"][ref.repo]
    tag_collections := _collection_by_tag(full_collection, tag)
	items := time_lib.acceptable_items(full_collection)
} else := []

_collection_by_tag(all_items, tag) := items {
   items := [item |
     some item in items
     item.tag == tag
   ]
   count(items) > 0
} else := all_items
```

Reply:
Not sure about changing `_collection`, the `tag` then must be passed in somehow...

My brain is a bit slow today. I think it might cover (most/all?) cases if I add a `is_equal` variant that matches on both tag and digest.

Pushed [1fb7ef1](https://github.com/enterprise-contract/ec-policies/pull/787/commits/1fb7ef1831484b4f0fdc298a6811021b501f1081) to demonstrate the effects.

I'm not 100% sure of this...

Reply:
> Not sure about changing _collection, the tag then must be passed in somehow...

Isn't the tag an attribute of `ref`, i.e. `ref.tag`? (I didn't do that in my example code above)

Reply:
The `is_equal` approach may work, but it may have its own challenges. For example, we want to take the tag into account only if at least one record in the collection has a matching tag. Otherwise, we want to consider the whole collection as is. This may be challenging to do when comparing each record in the collection individually.